### DISCUSSION TITLE
New notification logic

### DISCUSSION THREAD

**Discussion by BorisFromBE** [openkfw/TruBudget#1147] [Ideas]
Local business processes in Burkina Faso require a new logic for notifications. In order to make the behavior of those notifications compatible with those local processes, 4 needs have been identified :

1. Adapt the content of some existing notifications
2. Adapt some rules about the recipient of the notification
3. Adapt the triggers of some notifications 
4. Add new notifications corresponding to additional events

![Image1](https://user-images.githubusercontent.com/91593570/168257673-05ec679a-d887-4766-b235-0e224a7654b7.png)

**Comment by Stezido** [comment]
The notification system in TruBudget is really simple and it cannot be modified as you described.
The problems the new notification system has to solve are: 

- set different content of a notification
- set recipients of a notification
- set new triggers on events

**set different content**
This is tricky because it's a organization specific requirement but it's currently part of TruBudget core. Considering the notifications may not be shared via blockchain a notification-microservice can be created where all notifications are handled and can be adapted for an organization. This means notifications can still be part of TruBudget's core components via an extra microservice (blockchain, api, notification-service, frontend) but can be adjusted by an organization via a dashboard.

**set recipients**
One solution is a system like github's notification system where a user can un-/subscribe to a certain event. 

**set new triggers on events**
The events of TruBudget are fixed anyway so being able to add new triggers is not necessary, but TruBudget should provide a trigger on each existing event of TruBudget. The multichain-feed go script is a current solution to execute code depending on a blockchain transaction which is always related to an event. Before implementing a new notification system which should react on all events of TruBudget, an event trigger system like the multichain-feed must be adapted and maybe reworked in terms of scalability and efficiency.

**Reply by georgimld** [reply]
After some brainstorming about how the 3 features mentioned above could be implemented, we have the following ideas:

### Subscription zu einem Event - low hanging fruit - only api changes & a button in the UI

- Users could subscribe to a notification on a specific projekt/subproject/worklfowitem (per ressource we would save a subscription list)
- Subscribers will be saved on a stream - e.g. a new stream called _project_notifications_
- User needs to have the "View" permission on a specific ressource in order to see the notification
- important to check when the view rights were updated and when a notification was published - users can only see the notifications that were created at a time where they also had view permissions (double eventsourcing when reading)
- notification_create -> subscribers are saved as recipients (currently only assignee - future all subscribers)
- notifications will be placed in the correct stream in the future (e.g. _project_notifications_, etc)
- the api also has to read from the old notifications stream (for backwards-compatibility)


### Configuring content of notification - not a low hanging fruit
-  updating the content of a notification could be done in 2 ways:
    - over a config-file with predefined messages
    - over a dashboard, where the user can configure the notification messages - can then be saved as a separate event in the _project_notifications_ stream
- difficulties: how to tackle language files, placeholders, and so on

### Trigger of notifications - low hanging, could be customized individually for specifc use case
- the go script could be adjusted directly to listen to specific events & what should happen with these notifications (e.g. send emails, notify other systems, so on)
- the multichain-feed documentation already explains how to modify the events to listen to: https://github.com/openkfw/TruBudget/tree/main/multichain-feed

**Comment by LukasKM** [comment]
The item 8 in the list currently works in a way that when a user is assigned a task/WF item, he is notified. However, in practice there can be a considerable time difference between being assigned a task, and starting to work on it. Therefore there should ideally be two additional notifications sent: 
1) when the previous task has been validated, i.e. to say that you can work on the task because it is your turn. 
2) if there is any deadline/due date set for this task, the user should be notified that the due date is approaching. The notification could be sent based on a defined parameter (set up on node level), e.g. 2 weeks before, and/or also on the due date.
